rust-symcrypt/src/hmac.rs

```diff
@@ -1,180 +1,419 @@
+/* Hmac functions. For further documentation please refer to symcrypt.h */
+
+use crate::errors::SymCryptError;
 use core::ffi::c_void;
 use std::mem;
+use std::pin::Pin;
 use std::ptr;
+use std::sync::Arc;
 use symcrypt_sys;
 
 pub const SHA256_HMAC_RESULT_SIZE: usize = symcrypt_sys::SYMCRYPT_SHA256_RESULT_SIZE as usize;
 pub const SHA384_HMAC_RESULT_SIZE: usize = symcrypt_sys::SYMCRYPT_SHA384_RESULT_SIZE as usize;
 
+/// Generic trait for stateful Hmac functions
+///
+/// append() appends data to the HmacShaXXXState, this operation can be done multiple times.
+///
+/// result() returns the result of the Hmac. Once result() is called, the lifetime of the HmacShaXXXState is
+/// finished and HmacShaXXState will be drop()'d. To perform other stateful hash operation you must create
+/// a new hash object via HmacShaXXXState::new()
+///
+/// copy() creates a copy of the current HmacShaXXXState. copy() will create a new state but will reference the same
+/// expanded_key of the current HmacShaXXXState; therefore increasing the refcount of the HmacShaXXXExpandedKey struct.
+pub trait HmacState {
+    type Result;
+
+    fn append(&mut self, data: &[u8]);
+    fn result(self) -> Self::Result;
+    fn copy(&self) -> Box<Self>;
+}
+
+/// Wrapper for the expanded key that is Pin<Box<>>'d because we want to ensure that its memory address is heap allocated and not moved through its lifetime.
+struct HmacSha256ExpandedKey(Pin<Box<symcrypt_sys::SYMCRYPT_HMAC_SHA256_EXPANDED_KEY>>);
+
+/// Since HmacSha256ExpandedKey can be referenced multiple times, HmacSha256ExpandedKey must be ref counted and a there needs to be a separate drop()
+impl Drop for HmacSha256ExpandedKey {
+    fn drop(&mut self) {
+        unsafe {
+            // SAFETY: FFI calls
+            symcrypt_sys::SymCryptWipe(
+                ptr::addr_of_mut!(*self.0) as *mut c_void, // * to dereference the Box<>
+                mem::size_of_val(&mut *self.0) as symcrypt_sys::SIZE_T, // * to dereference the Box<>
+            );
+        }
+    }
+}
+
+/// Using an inner HmacSha256 state that is Pin<Box<>>'d since the memory address for Self is moved around when returning from HmacSha256State::new()
+///
+/// expanded_key is Arc<>'d to be properly ref counted when calling HmacShaXXXState::copy().
+///
+/// SymCrypt expects the address for its structs to stay static through the structs lifetime to guarantee that structs are not memcpy'd as
+/// doing so would lead to use-after-free and inconsistent states.
 pub struct HmacSha256State {
+    inner: Pin<Box<HmacSha256Inner>>,
+}
+
+struct HmacSha256Inner {
     state: symcrypt_sys::SYMCRYPT_HMAC_SHA256_STATE,
-    expanded_key: symcrypt_sys::SYMCRYPT_HMAC_SHA256_EXPANDED_KEY,
+    expanded_key: Arc<HmacSha256ExpandedKey>,
+}
+
+unsafe impl Send for HmacSha256Inner {
```

Comment:
Based on the rustls `Send + Sync` parameters on their traits. We will follow up with them offline to see their requirements.  I think that async can get a little dicey here so we'd like some more context before implementing.

Reply:
For this impl to be safe, you'll need to make sure SymCrypt itself is thread-safe. Alternatively, you can wrap the `HmacSha256Inner` field in `HmacSha256State` with a `Mutex` and then `HmacSha256State` should automatically be `Send` without needing any `unsafe impl`s.

Reply:
Let's first talk to the rustls folks before adding a Mutex.

I'm not seeing other implementations like ring not adding a Mutex.